src/service/payment.service.ts

```diff
@@ -0,0 +1,87 @@
+  req: Request, session: Session, transactionId: string, overseasEntityId: string, transactionRes
+): Promise<string> => {
+
+  const paymentUrl = transactionRes.headers?.[PAYMENT_REQUIRED_HEADER];
+
+  // If the transaction response is fee-bearing, a `X-Payment-Required` header will be received,
+  // directing the application to the Payment Platform to begin a payment session
+  if (isActiveFeature(FEATURE_FLAG_ENABLE_PAYMENT_16052022) && paymentUrl) {
+    const createPaymentRequest: CreatePaymentRequest = setPaymentRequest(transactionId, overseasEntityId);
+
+    // Save info into the session extra data field, including the state used as `nonce` against CSRF.
+    setApplicationData(session, { ...createPaymentRequest, transactionId, overseasEntityId }, PaymentKey);
+
+    // Create Payment Api Client by using the `paymentUrl` as baseURL
+    const apiClient: ApiClient = createOAuthApiClient(session, paymentUrl);
+
+    // Calls the platform to create a payment session
+    const paymentResult = await apiClient.payment.createPaymentWithFullUrl(createPaymentRequest);
+
+    // Verify the state of the payment, success or failure (eg. cost not found, connection issues ...)
+    if (paymentResult.isFailure()) {
+      const errorResponse = paymentResult.value;
+
+      const msgErrorStatusCode = `http response status code=${ errorResponse?.httpStatusCode || "No Status Code found in response" }`;
+      const msgErrorResponse = `http errors response=${ JSON.stringify(errorResponse?.errors || "No Errors found in response") }`;
+      const msgError = `payment.service failure to create payment, ${msgErrorStatusCode}, ${msgErrorResponse}.`;
+
+      throw createAndLogErrorRequest(req, msgError);
+    } else if (!paymentResult.value?.resource) {
+      throw createAndLogErrorRequest(req, "No resource in payment response");
+    } else {
+      const paymentResource: Payment = paymentResult.value.resource;
+
+      logger.infoRequest(req, `Create payment, status_code=${ paymentResult.value.httpStatusCode }, status=${ paymentResource.status }, links= ${ JSON.stringify(paymentResource.links ) } `);
+
+      // To initiate the web journey through which the user will interact to make the payment
+      return paymentResource.links.journey;
+    }
+  } else {
+    // Only if transaction does not have a fee or PAYMENT flagged.
+    return CONFIRMATION_URL;
+  }
+};
+
+const setPaymentRequest = (transactionId, overseasEntityId): CreatePaymentRequest => {
+
+  const paymentResourceUri = `${API_URL}/transactions/${transactionId}/${PAYMENT}`;
+
+  const baseURL = `${CHS_URL}${REGISTER_AN_OVERSEAS_ENTITY_URL}`;
+  const reference = `${REFERENCE}_${transactionId}`;
+
+  // Once payment has been taken, the platform redirects the user back to the application,
+  // using the application supplied `redirectUri`.
+  const redirectUri = `${baseURL}${TRANSACTION}/${transactionId}/${OVERSEAS_ENTITY}/${overseasEntityId}/${PAYMENT}`;
```

Comment:
It looks like this URL closely matches the URL you defined in your new constant above - PAYMENT_WITH_TRANSACTION_URL - albeit with substituted values. Just wondering, rather than repeat the format of it here, could you somehow reference and use that constant and substitute the transaction and OE ids in it (because if the format did ever change, it would need updating in two places)?

Reply:
Hi Martin, could you please be more clear, I do not see any URL closely match

Reply:
So that constant is built up from:

```
const TRANSACTION_PATH = TRANSACTION + "/:transactionId/";
const OVERSEAS_ENTITY_PATH = OVERSEAS_ENTITY + "/:overseaEntityId/";
export const PAYMENT_WITH_TRANSACTION_URL = REGISTER_AN_OVERSEAS_ENTITY_URL + TRANSACTION_PATH + OVERSEAS_ENTITY_PATH + PAYMENT;
```

Which I think resolves to: '**/register-an-overseas-entity/transaction/:transactionId/overseas-entity/:overseasEntityId/payment**'

So it looks like it's the same format as the '**redirectUri**' you build.

Give me a call if still unclear and we can have a chat...

Reply:
The `PAYMENT_WITH_TRANSACTION_URL` is the pattern used for our endpoint, `redirectUri` will have instead the field correctly populated with the `transactionId` and `overseaEntityId` and has to match the endpoint.
The first is always the same, the second is changing every time.

Lets have a call, you know I love them :D